### PR TITLE
Added GemJoin8 and GUSD example

### PR DESCRIPTION
**Pull request by hexonaut**
Same code adapted from here: https://github.com/makerdao/dss-deploy/pull/56

Added GemJoin8 to support GUSD. To improve on re-usability for proxy contracts, I've allowed for arbitrary function names which return the current implementation.

### CODE REVIEW

**Comment by hexonaut** [comment]
I'm going back to a static call for the implementation due to gas cost concerns. Having a static call to `gem.erc20Impl()` saves over 1000 gas which is high enough to be worthwhile imo. Here are the tests I ran:

```
Dynamic (No opt) = 75684
Dynamic (Opt, Runs=1) = 75572
Dynamic (Opt, Runs=200) = 75313

Static (No opt) = 74513
Static (Opt, Runs=1) = 74389
Static (Opt, Runs=200) = 74130
```

File: src/join-8.sol
Comment (godsflaw) [inline src/join-8.sol:54]: you can yank this comment now.

**Review by godsflaw — APPROVED** [review]
LGTM after comment removal

**Comment by godsflaw** [comment]
It occurred to me this is basically the same as the Tether adapter, but with the proper transfer checks.  If you've not compared the two, you should diff those two for one last comparison.

**Comment by hexonaut** [comment]
Yeah they are roughly the same. I copied in the code from join-6 for the implementation checks.

**Comment by KentonPrescott** [comment]
Hi @hexonaut , before merging, could you confirm the completed items on the [Contributor Checklist](https://github.com/makerdao/dss-gem-joins/blob/master/.github/PULL_REQUEST_TEMPLATE.md#contribution-checklist)?

File: src/join-8.sol
Comment (brianmcmichael) [inline src/join-8.sol:92]: Minor thing, but can you add a newline at the end of the file here.

**Comment by hexonaut** [comment]
Adding contributor checklist because I opened this before the template existed.

# Description

Collateral Type (e.g. `ETH-A`): GUSD-A

COB Team Name or Author(s): @hexonaut @godsflaw 

GemJoin Adapter: GemJoin8

Is this GemJoin adapter new? : Yes

If so, why is a new one needed? : There is no existing adapter that checks for implementation using `erc20Impl()` as the view.


# Contribution Checklist

- [ ] Collateral token contract has been added in `src/tokens/` as `TOKEN.sol`, if not already present
- [ ] If needed, a new `GemJoin` has been added in `src/` as `join-X.sol`
- [ ] A new test (`testGemJoinX_TOKEN()`) has been added to `src/join.t.sol` and follows previous tests as examples
- [ ] If needed, other tests have been added to verify functionality of the new `GemJoin`:
    - [ ] `testFailGemJoinXJoinWad()`
    - [ ] `testFailGemJoinXExitWad()`
    - [ ] `testFailGemJoinXJoin()`
    - [ ] `testFailGemJoinXExit()`
    - [ ] `testFailJoinAfterCageGemJoinX()`
- [ ] README has been updated to reflect a new token and `GemJoin` addition

# Checklist

- [ ] Confirm that token contract matches the verified source on etherscan mainnet
- [ ] If a new `GemJoin` is added, ensure the above description is accurate
- [ ] If a new `GemJoin` is added, ensure a new `GemJoin` has been added with the correct name
- [ ] Ensure README is updated
- [ ] Manually review the new test(s)
- [ ] Manually run dapp tests in `src/join.t.sol` to confirm they all pass

File: src/join-8.sol
Comment (brianmcmichael) [inline src/join-8.sol:48]: Add a space here just to line up with the others.

File: src/join.t.sol
Comment (brianmcmichael) [inline src/join.t.sol:461]: use `uint256` here for consistency

File: src/join.t.sol
Comment (brianmcmichael) [inline src/join.t.sol:588]: `uint256`

File: src/join.t.sol
Comment (brianmcmichael) [inline src/join.t.sol:616]: I think these `Fail` tests are failing on the previous line.

The function signature of the setter is `gusd.setImplementation(address, uint256);`, so these tests would fail on the implementation set, rather than the join.

Comment (brianmcmichael) [reply src/join.t.sol:616]: Ah, nevermind, the test GUSD is using a simplified setter. Disregard.

Comment (hexonaut) [reply src/join.t.sol:616]: This is setting the implementation on the `gem` instead of the `gemJoin`. They just happen to have the same method name. It should be correct behaviour.

**Comment by brianmcmichael** [comment]
Functionally seems good. Just a few minor code formatting nits, also be sure to check all of the `uint` usages and update them to `uint256`

File: src/join.t.sol
Comment (hexonaut) [reply src/join.t.sol:616]: Just double checked and commenting out the failure line results in the test passing, so it is correct.

**Comment by hexonaut** [comment]
Changes done. All tests passed:

```
Running 40 tests for src/join.t.sol:DssDeployTest
[PASS] testGemJoin_PAXUSD() (gas: 40823422)
[PASS] testGemJoin_REP() (gas: 40823466)
[PASS] testFailJoinAfterCageGemJoin4() (gas: 41315613)
[PASS] testFailJoinAfterCageGemJoin3() (gas: 40728698)
[PASS] testGemJoin5_WBTC() (gas: 40796066)
[PASS] testGemJoin_LINK() (gas: 40828178)
[PASS] testFailTokenSai() (gas: 41770263)
[PASS] testFailGemJoin7Exit() (gas: 41471143)
[PASS] testGemJoin_BAL() (gas: 40828177)
[PASS] testFailJoinAfterCageGemJoin8() (gas: 40960981)
[PASS] testGemJoin_BAT() (gas: 40823509)
[PASS] testGemJoin2_OMG() (gas: 41188581)
[PASS] testGemJoin8_GUSD() (gas: 41115587)
[PASS] testGemJoin_KNC() (gas: 40823488)
[PASS] testFailJoinAfterCageGemJoin7() (gas: 41448734)
[PASS] testGemJoin3_DGD() (gas: 40774732)
[PASS] testGemJoin7_USDT() (gas: 41500568)
[PASS] testGemJoin4_GNT() (gas: 41333055)
[PASS] testFailGemJoin7JoinWad() (gas: 272747604447804)
[PASS] testGemJoin5_USDC() (gas: 40796064)
[PASS] testFailJoinAfterCageAuthGemJoin() (gas: 41839721)
[PASS] testFailGemJoin7Join() (gas: 41399775)
[PASS] testGemJoin_YFI() (gas: 40828156)
[PASS] testFailGemJoin6Exit() (gas: 41115623)
[PASS] testFailGemJoin8Exit() (gas: 40960937)
[PASS] testFailJoinAfterCageGemJoin6() (gas: 41116190)
[PASS] testFailGemJoin8Join() (gas: 40974300)
[PASS] testFailGemJoin6Join() (gas: 41052709)
[PASS] testFailGemJoin7ExitWad() (gas: 41451116)
[PASS] testGemJoin_ZRX() (gas: 40823442)
[PASS] testGemJoin6_DGD() (gas: 41165757)
[PASS] testGemJoin7FeeChange() (gas: 41556323)
[PASS] testTokenSai() (gas: 41906587)
[PASS] testFailGemJoin8JoinWad() (gas: 40960937)
[PASS] testFailGemJoin8ExitWad() (gas: 40960981)
[PASS] testFailJoinAfterCageGemJoin2() (gas: 41136297)
[PASS] testFailJoinAfterCageGemJoin5() (gas: 40749853)
[PASS] testGemJoin_MANA() (gas: 40828198)
[PASS] testGemJoin_COMP() (gas: 40828153)
[PASS] testGemJoin_LRC() (gas: 40828219)
```

**Review by brianmcmichael — APPROVED** [review]
Looks good.

**Pull request merged**
